// @tldraw/store/dist-esm/index.mjs
import { registerTldrawLibraryVersion } from "@tldraw/utils";
import { AtomMap } from "./lib/AtomMap.mjs";
import { devFreeze } from "./lib/devFreeze.mjs";
import { IncrementalSetConstructor } from "./lib/IncrementalSetConstructor.mjs";
import {
  createMigrationIds,
  createMigrationSequence,
  createRecordMigrationSequence,
  defineMigrations,
  MigrationFailureReason,
  parseMigrationId
} from "./lib/migrate.mjs";
import {
  createEmptyRecordsDiff,
  isRecordsDiffEmpty,
  reverseRecordsDiff,
  squashRecordDiffs,
  squashRecordDiffsMutable
} from "./lib/RecordsDiff.mjs";
import { assertIdType, createRecordType, RecordType } from "./lib/RecordType.mjs";
import {
  createComputedCache,
  Store
} from "./lib/Store.mjs";
import { StoreQueries } from "./lib/StoreQueries.mjs";
import { StoreSchema } from "./lib/StoreSchema.mjs";
import {
  StoreSideEffects
} from "./lib/StoreSideEffects.mjs";
registerTldrawLibraryVersion(
  "@tldraw/store",
  "3.15.6",
  "esm"
);
export {
  AtomMap,
  IncrementalSetConstructor,
  MigrationFailureReason,
  RecordType,
  Store,
  StoreQueries,
  StoreSchema,
  StoreSideEffects,
  assertIdType,
  createComputedCache,
  createEmptyRecordsDiff,
  createMigrationIds,
  createMigrationSequence,
  createRecordMigrationSequence,
  createRecordType,
  defineMigrations,
  devFreeze,
  isRecordsDiffEmpty,
  parseMigrationId,
  reverseRecordsDiff,
  squashRecordDiffs,
  squashRecordDiffsMutable
};
