// tldraw/dist-cjs/lib/tools/SelectTool/childStates/Crop/children/PointingCrop.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PointingCrop_exports = {};
__export(PointingCrop_exports, {
  PointingCrop: () => PointingCrop
});
module.exports = __toCommonJS(PointingCrop_exports);
var import_editor = require("@tldraw/editor");
class PointingCrop extends import_editor.StateNode {
  static id = "pointing_crop";
  onCancel() {
    this.editor.setCurrentTool("select.crop.idle", {});
  }
  onPointerMove(info) {
    if (this.editor.inputs.isDragging) {
      this.startDragging(info);
    }
  }
  onLongPress(info) {
    this.startDragging(info);
  }
  onPointerUp(info) {
    this.editor.setCurrentTool("select.crop.idle", info);
  }
  startDragging(info) {
    this.editor.setCurrentTool("select.crop.translating_crop", info);
  }
}
