// @tldraw/state-react/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  track: () => import_track.track,
  useAtom: () => import_useAtom.useAtom,
  useComputed: () => import_useComputed.useComputed,
  useQuickReactor: () => import_useQuickReactor.useQuickReactor,
  useReactor: () => import_useReactor.useReactor,
  useStateTracking: () => import_useStateTracking.useStateTracking,
  useValue: () => import_useValue.useValue
});
module.exports = __toCommonJS(src_exports);
var import_utils = require("@tldraw/utils");
var import_track = require("./lib/track");
var import_useAtom = require("./lib/useAtom");
var import_useComputed = require("./lib/useComputed");
var import_useQuickReactor = require("./lib/useQuickReactor");
var import_useReactor = require("./lib/useReactor");
var import_useStateTracking = require("./lib/useStateTracking");
var import_useValue = require("./lib/useValue");
(0, import_utils.registerTldrawLibraryVersion)(
  "@tldraw/state-react",
  "3.7.0",
  "cjs"
);
